[docs] Follower reads in YSQL (#10670)
yugabyte/yugabyte-db: docs/content/latest/explore/multi-region-deployments/follower-reads-ysql.md

  latest:
    identifier: explore-multi-region-deployments-follower-reads-ysql
    parent: explore-multi-region-deployments
    weight: 280
isTocNested: true
showAsideToc: true
---

<ul class="nav nav-tabs-alt nav-tabs-yb">
  <li >
    <a href="../follower-reads-ysql/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>YSQL</a>
  </li>

  <li >
    <a href="../read-replicas-ycql/" class="nav-link">
      <i class="icon-cassandra" aria-hidden="true"></i>YCQL</a>
  </li>
</ul>

## Introduction

> YugabyteDB requires reading from the **leader** to read the latest data. However, for applications that do not require the latest data, and/or are working with unchanging data; the cost of contacting a potentially remote leader to fetch the data may be wasteful. The application could benefit with better latency by reading from a **replica** that may be closer to the client.

Reviewer: `benefit with` sounds off. Perhaps a native speaker may check if this is the right usage.

Since the replicas may not be up to date with respect to all the updates, it should be noted that this design may respond with stale data. The user may specify how much staleness the application can tolerate. When enabled, read-only operations may be handled by the closest replica; instead of having to go to the leader.

## Surface area and Usage

### Surface area

> - *yb_read_from_followers* : This session variable controls whether read from followers is enabled.
> 
> - *yb_follower_read_staleness_ms* : This session variable defines the maximum staleness that the user is willing to tolerate.
> 
> ### Default behavior
> 
> - *yb_read_from_followers* : Defaults to false.
> 
> - *yb_follower_read_staleness_ms* : Defaults to 30 seconds.

Reviewer 2: ```suggestion
Two session variables control the behavior of follower reads:

- `yb_read_from_followers` controls whether reading from followers is enabled. Default is false.

- `yb_follower_read_staleness_ms` sets the maximum allowable staleness. Default is 30 seconds.
```

### Expected behavior

The table describes what the expected behavior is when a read happens from a follower.

| Conditions | Expected behavior |
| --- | --- |

> | yb_read_from_followers is true AND <br> (Transaction is marked read-only <br>OR<br> Single-statement-select can be inferred to be read-only) <br> | Read will happen from the closest follower|

Reviewer: single statement select inferrence isn't landed yet.

Reviewer 2: My suggestion below addresses this, and we should add a separate row to the table when it does land. There's no need for complex logic in a reference table. :)

> | Conditions | Expected behavior |
> | --- | --- |
> | yb_read_from_followers is true AND <br> (Transaction is marked read-only <br>OR<br> Single-statement-select can be inferred to be read-only) <br> | Read will happen from the closest follower|

Reviewer 2: ```suggestion
| Conditions | Expected behavior |
| :--------- | :---------------- |
| yb_read_from_followers is true AND transaction is marked read-only | Read happens from the closest follower |
```

> |yb_read_from_followers is false OR <br> Transaction/statement is not read only | Read will happen from the leader |

Reviewer 2: ```suggestion
| yb_read_from_followers is false OR transaction/statement is not read-only | Read happens from the leader |
```

### Read from follower conditions

> - If the follower’s safe-time is larger than the read-time thus chosen, the follower may serve the read without any delay.

Reviewer 3: "thus chosen" -- chosen how? This hasn't been explained before on this page.

Reviewer 2: Do we actually mean "If the follower's data is fresher than the client's yb_follower_read_staleness_ms value" here?

Reviewer: I'd say rewite this as:

If the follower's safe-time is at least <current_time> - <staleness>, 

If the follower is not yet caught up to <current_time> - <staleness> ...

> - If the follower is not yet caught up to the specified timestamp, the read will be redirected to a different replica transparently from the end-user. The end user may see a slight increase in latency depending on the location of the replica which satisfies the read.

Reviewer 3: "the specified timestamp" -- which timestamp? Again, this concept has not been explained.

Reviewer: If the follower is not yet caught up to `<current_time> - <staleness> `...

>   - The current behavior is that a follower/replica would wait for the requested timestamp to become safe. This behavior shall be changed to respond back to the `ybclient/pggate` immediately, so that the request may be routed to a different server that has the data.

Reviewer 3: This discussion is not relevant to external documentation, @amitanandaiyer .

Reviewer: Yes. The 3 sub bullets:
![Screen Shot 2021-11-30 at 10 32 08 AM](https://user-images.githubusercontent.com/32495790/144106759-599423cc-b9de-46e2-a621-409d192e6339.png)

may be removed from the external documentation.

  - The current redirection policy is to directly go to the leader if the follower read is rejected by a follower. This behavior could be also be changed going forward, to explore other replicas before going to the leader.
  - If the request reaches the leader, it will no longer reject/redirect the request.

### Read-only transaction conditions

Regarding marking the transaction as read only, a user can do one of the following:

> 

Reviewer 3: How come there's no mention of pg_hint syntax in this section as another way of achieving same thing?

Reviewer: updated the FS.

Can you pls also add

* SET transaction_read_only = TRUE (applies to the current transaction block)
* /*+ Set(transaction_read_only true) */ SELECT (applies only to the current transaction/select statement)
       
here.

> - `SET TRANSACTION READ ONLY` : Applies only to the current transaction block.

Reviewer 3: Minor: No space before the : 

Ditto on the next two lines also.

Reviewer 2: Even better: no colon, lowercase the word that follows: "`SET TRANSACTION READ ONLY` applies only to the..."

> - `SET TRANSACTION READ ONLY` : Applies only to the current transaction block.

Reviewer 2: ```suggestion
- `SET TRANSACTION READ ONLY` applies only to the current transaction block.
```

> - `SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY` : Applies the read-only setting for all statements and transaction blocks that follow.
> - `SET default_transaction_read_only = TRUE` : Applies the read-only setting for all statements and transaction blocks that follow.

Reviewer 2: ```suggestion
- `SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY` applies the read-only setting to all statements and transaction blocks that follow.
- `SET default_transaction_read_only = TRUE` applies the read-only setting to all statements and transaction blocks that follow.
```

## Examples

This example uses follower reads since the **transaction** is marked read-only.

```sql
set yb_read_from_followers = true;
start transaction read only;
SELECT * from t WHERE k='k1';
commit;
```

```output
 k  | v
----+----
 k1 | v1
```

This example uses follower reads since the **session** is marked read only.

```sql
set session characteristics as transaction read only;
set yb_read_from_followers = true;
SELECT * from t WHERE k='k1';
```

```output
 k  | v
----+----
 k1 | v1
(1 row)
```

> This next two examples use follower reads since **pg_hint** can be used during PREPARE and CREATE FUNCTION to do follower reads.

Reviewer 3: Can't pg_hint method be used for a top-level/standalone SELECT statement? Does it have to be only via PREPARE or CREATE FUNCTIONA?

Reviewer: Yes. the pg_hint method can also be used for stand_alone statements.

The important point to highlight there is that for Prepare/Functions, the pg_hint needs to be applied at the prepare/function-definition stage. Not at the execute stage.

Author: I will include it as a note.

```sql
set session characteristics as transaction read write;
set yb_read_from_followers = true;

PREPARE select_stmt(text) AS

> /*+ Set(transaction_read_only on)*/

Reviewer: Do we need a space before */

set yb_read_from_followers = true;

CREATE FUNCTION func() RETURNS text AS
$$ /*+ Set(transaction_read_only on)*/
SELECT * from t WHERE k=1 $$ LANGUAGE SQL;
CREATE FUNCTION
SELECT func();
```

```output
 k  | v
----+----
 k1 | v1
(1 row)
```

A join example that uses follower reads.

```sql
set yb_read_from_followers = true;
set session characteristics as transaction read only;
SELECT * from t1, t2 WHERE t1.k='k3' AND t1.v = t2.v;
```

```output
 k | v
---+---
(0 rows)
```

The following examples demonstrate **staleness** after enabling the `yb_follower_read_staleness_ms` property.

```sql
set session characteristics as transaction read write;
select * from t where k = 'k1';
```

```output
 k  | v
----+----
 k1 | v1
(1 row)
```

```sql
UPDATE t SET  v = 'v1+1' where k = 'k1';
/* sleep 10s */

UPDATE t SET  v = 'v1+2' where k = 'k1';
/* sleep 10s */

select * from t where k = 'k1';
```

```output
 k  |  v
----+------
 k1 | v1+2
(1 row)
```

```sql
set session characteristics as transaction read only;
set yb_read_from_followers = true;

select * from t where k = 'k1';
```

```output
 k  |  v
----+------
 k1 | v1
(1 row)
```

```sql
set yb_follower_read_staleness_ms = 5000;

> select * from t where k = 'k1';   /* 5s old value */

Reviewer 3: Is this 5s old value or up to 5s old value?

Author: Will change this to "up to 5s old value" wherever applicable.

Reviewer 2: (Same for the 15 and 25 versions :) )

```

```output
 k  |  v
----+------
 k1 | v1+2
(1 row)
```

```sql
set yb_follower_read_staleness_ms = 15000;

select * from t where k = 'k1';   /* 15s old value */
```

```output
 k  |  v
----+------
 k1 | v1+1
(1 row)
```

```sql
postgres=# set yb_follower_read_staleness_ms = 25000;

> postgres=# select * from t where k = 'k1';   /* 25s old value */

Reviewer 2: Why are these Postgres prompts and not ysqlsh?

```

```output
 k  |  v
----+------
 k1 | v1
(1 row)
```

> The final example demonstrates that SELECTS error out in read-only transactions.

Reviewer 3: The wording needs to be improved here. 

It is not that SELECTs error out in a read-only transaction. It has more do with the fact that SELECT INTO is not a read-only transaction.

And this example has nothing to do with follower reads. Because we would get the same error even if follower reads was off.

So what's the point  of this example?

CC: @amitanandaiyer , @aishwarya24 ?

Reviewer: Yes. That's true. We could just remove this example.

It was added to the FS to clarify the discussion regarding how we'd prevent/safegurard against a select statement  having non-read-only side-effects. 

Doesn't need to be part of the user documentation.

<ul class="nav nav-tabs-alt nav-tabs-yb">
  <li >
    <a href="../follower-reads-ysql/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>YSQL</a>
  </li>

  <li >
    <a href="../read-replicas-ycql/" class="nav-link">
      <i class="icon-cassandra" aria-hidden="true"></i>YCQL</a>
  </li>
</ul>

## Introduction

YugabyteDB requires reading from the **leader** to read the latest data. However, for applications that don't require the latest data, or are working with unchanging data, the cost of contacting a potentially remote leader to fetch the data may be wasteful. Your application may benefit from better latency by reading from a **replica** that is closer to the client.

Replicas may not be completely up to date with all updates, so this design may respond with stale data. You can specify how much staleness the application can tolerate. When enabled, read-only operations may be handled by the closest replica, instead of having to go to the leader.

## Surface area and usage

### Surface area

Two session variables control the behavior of follower reads:

- `yb_read_from_followers` controls whether reading from followers is enabled. Default is false.

> - `yb_follower_read_staleness_ms` sets the maximum allowable staleness. Default is 30 seconds.

Reviewer 3: is 30 seconds.
-->
is 30000 (i.e. 30 secs).

### Expected behavior

The table describes what the expected behavior is when a read happens from a follower.

| Conditions | Expected behavior |
| :--------- | :---------------- |
| yb_read_from_followers is true AND transaction is marked read-only | Read happens from the closest follower |
| yb_read_from_followers is false OR transaction/statement is not read-only | Read happens from the leader |

### Read from follower conditions

- If the follower’s safe-time is at least `<current_time> - <staleness>`, then the follower may serve the read without any delay.

- If the follower is not yet caught up to `<current_time> - <staleness>`, then the read will be redirected to a different replica transparently from the end-user. The end user may see a slight increase in latency depending on the location of the replica which satisfies the read.

### Read-only transaction conditions

> Regarding marking the transaction as read only, a user can do one of the following:

Reviewer 3: Regarding marking the transaction
-->
To mark a transaction

- `SET TRANSACTION READ ONLY` applies only to the current transaction block.
- `SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY` applies the read-only setting to all statements and transaction blocks that follow.
- `SET default_transaction_read_only = TRUE` applies the read-only setting to all statements and transaction blocks that follow.

> - `/*+ Set(transaction_read_only true) */ SELECT` applies only to the current transaction/select statement.

Reviewer 3: Use [pg_hint_plan](https://docs.yugabyte.com/latest/explore/query-1-performance/pg-hint-plan/) mechanism	to embed the hint along with the SELECT statement. For example,` /*+ Set(transaction_read_only true) */ SELECT ...`  applies only	to the current SELECT  statement.

## Examples

This example uses follower reads since the **transaction** is marked read-only.

```sql
set yb_read_from_followers = true;
start transaction read only;
SELECT * from t WHERE k='k1';
commit;
```

```output
 k  | v
----+----
 k1 | v1
```

This example uses follower reads since the **session** is marked read only.

```sql
set session characteristics as transaction read only;
set yb_read_from_followers = true;
SELECT * from t WHERE k='k1';
```

```output
 k  | v
----+----
 k1 | v1
(1 row)
```

> The following examples use follower reads since **pg_hint** can be used during PREPARE, CREATE FUNCTION, and SELECT to do follower reads.

Reviewer 3: replace all occurrences of pg_hint with pg_hint_plan, as I believe that's the correct word.

since **pg_hint**
-->
since the **pg_hint_plan** mechanism

{{< note title="Note" >}}

> The pg_hint needs to be applied at the prepare/function-definition stage and not at the `execute` stage.

Reviewer 3: The pg_hint needs to 
-->
The pg_hint_plan hint needs to

{{< /note >}}

```sql
set session characteristics as transaction read write;
set yb_read_from_followers = true;

PREPARE select_stmt(text) AS
/*+ Set(transaction_read_only on) */
SELECT * from t WHERE k=$1;

EXECUTE select_stmt(‘k1’);
```

```output
 k  | v
----+----
 k1 | v1
(1 row)
```

```sql
set session characteristics as transaction read write;
set yb_read_from_followers = true;

CREATE FUNCTION func() RETURNS text AS
$$ /*+ Set(transaction_read_only on) */
SELECT * from t WHERE k=1 $$ LANGUAGE SQL;
CREATE FUNCTION
SELECT func();
```

```output
 k  | v
----+----
 k1 | v1
(1 row)
```

```sql

> set session characteristics as transaction read write;

Reviewer 3: Why is this line needed?

And overall the simple SELECT with the embedded hint should be the first example for pg_hint_plan, instead of the more uncommon cases of PREPARE and CREATE FUNCTION. The flow needs to be simple to advanced rather than reverse.

Reviewer: It is just to highlight that pg_hint can be used in a read_write session to override the read only property.

We don't need this line per se. The example would work even without this line. So we can remove this, and otehr places in line 125 and 107.

Also, agree with moving this example above line 107. We can show this example, then the PREPARE example and then the CREATE FUNCTION example.

Author: Thank you ! I'll make the change accordingly.

set yb_read_from_followers = true;
/*+ Set(transaction_read_only on) */
SELECT * from t WHERE k='k1';
```

```output
----+----
 k1 | v1
(1 row)
```

A **join** example that uses follower reads.

```sql
create table table1(k int primary key, v int);
create table table2(k int primary key, v int);
insert into table1 values (1, 2), (2, 4), (3, 6), (4,8);
insert into table2 values (1, 3), (2, 6), (3, 9), (4,12);
set yb_read_from_followers = true;
set session characteristics as transaction read only;
select * from table1, table2 where table1.k = 3 and table2.v = table3.v;
```

```output
 k | v | k | v
---+---+---+---
 3 | 6 | 2 | 6
(1 row)
```

The following examples demonstrate **staleness** after enabling the `yb_follower_read_staleness_ms` property.

```sql

> set session characteristics as transaction read write;

Reviewer 3: @amitanandaiyer , @aishwarya24 :

between 'v1' becoming 'v1+1' don't you need a 10sec sleep to illustrate the point. I don't see that being explicitly mentioned. Would be good to show the INSERT of 'v1' and a sleep 10s after that also, in my opinion.

Reviewer: So the assumption here was that the initial state of the row was `v1` (as depicted by the select). With the write being made (at least 5+ seconds) in the past.

we could add it here, to make it more explicitly, if the default assumption is that the table/row is empty.

So replace the

```
insert into t values ('k1', 'v1')
/* sleep 10s */
```

select * from t where k = 'k1';
```

```output
 k  | v
----+----
 k1 | v1
(1 row)
```

```sql
UPDATE t SET  v = 'v1+1' where k = 'k1';
/* sleep 10s */

UPDATE t SET  v = 'v1+2' where k = 'k1';
/* sleep 10s */

select * from t where k = 'k1';
```

> ```output

Reviewer 3: Might be worth adding some commentary along with..

For example, here I would say:

This selects the latest version of the row because the transaction setting for the session is `read write`.

Two session variables control the behavior of follower reads:

- `yb_read_from_followers` controls whether reading from followers is enabled. Default is false.

> - `yb_follower_read_staleness_ms` sets the maximum allowable staleness. Default is 30000(30 seconds).

Reviewer 3: there should be a space before (
